### DISCUSSION TITLE
Phase field models of grain growth

### DISCUSSION THREAD

**Discussion by alikhan-zada** [idaholab/moose#17912] [Q&A Modules: Phase field] [answered]
Hello everyone,

**_Novice question alert!_** 🙃

I started studying the phase field method and there are some things that I am confused about.
So, if I have a multicomponent-multiphase material, say something like steel, would the standard grain growth model implemented in MOOSE be enough to visualize grain growth under a heat source, like in welding?

I am not interested in dendrite formation or the various phase transformations that might take place at different temperatures, but rather with the end microstructure of the steel.
Also, what other free energy components would I need to add in order to find out the material properties across the weld structure, for instance I would like to see how the stress and hardness varies across the weld as well as along it, in the various regions such as the HAZ, CGHAZ and FGHAZ?

Another thing that I would like to do is to know the different component segregation across the weld profile and if both, the component and the above part can be done together or not?

Thank you for reading my post, I would really appreciate your answers

Regards,
Ali

**Comment by laagesen** [comment, marked as answer]
Hi Ali- we have done something similar to what it sounds like you want to do. Please see the following paper:

https://www.tandfonline.com/doi/full/10.1080/00295450.2020.1838877

Here we used the standard grain growth model in MOOSE, although it is not really physically parameterized for a particular material system. We didn't really look at properties in this case- to do that, you would need to take the resulting microstructures and import them into a crystal plasticity model to see how they behave under deformation.

**Reply by alikhan-zada** [reply]
Thank you @laagesen, for your answer, but what I want to do is something similar to 
![Screenshot_1](https://user-images.githubusercontent.com/84128937/119851971-d9a63b80-bf49-11eb-9d76-7cb8dd5e5271.png)

Is it possible to have two decoupled domains where I can transfer the whole heat profile changing in time to another domain which solves only the phase field equation using MultiApps?

Or the other option is to just output the heat data over the domain w.r.t. space and time and just use that as an input to the second domain. 

Regards, 
Ali

**Reply by laagesen** [reply]
The first option would be doable via MultiApps. There are many different ways to transfer data among MultiApps. The documentation page below has a comprehensive list:

https://mooseframework.inl.gov/syntax/Transfers/

However before delving into this too deeply, I would try to figure out how you are going to model the heat transfer problem. Do you have a model for this in mind already?

**Reply by alikhan-zada** [reply]
The heat source model for a laser is already implemented in MOOSE: 

https://mooseframework.inl.gov/source/materials/FunctionPathEllipsoidHeatSource.html

I am thinking of changing the source term to that of a double ellipsoid model which should be fairly simple. The only thing that would be left would be to match the domain sizes and non-dimensionalize the parameters accordingly.

**Reply by SudiptaBiswas** [reply]
I did similar simulations by changing the GB Mobility as a function of time as the heat source moves. That is also another way to simplify the grain growth simulations.

**Reply by alikhan-zada** [reply]
@SudiptaBiswas, is it possible for you to share the input files?
Honestly, it would save me from doing a lot of things myself :)

**Reply by SudiptaBiswas** [reply]
Sorry, it involves a small development that was not checked in to MOOSE. I don't have an input file that will come in handy for you. You can design the mobility material following the heat source material you mentioned and provide the material as mobility for the grain growth kernels.

**Reply by alikhan-zada** [reply]
Hello again Larry and Sudipta, 
Thank you for your help so far.

After going through some literature, I have come across something that is very close to what I want to do (which was already in my initial question). 

So, I have two equations:
![image](https://user-images.githubusercontent.com/84128937/121287819-47287380-c91d-11eb-89ab-6b5e00ea9d3f.png)

The total free energy F is given as:
![image](https://user-images.githubusercontent.com/84128937/121287862-5b6c7080-c91d-11eb-8afe-0792fd8b3e0f.png)

The local free energy density consists of two parts and is given as:
![image](https://user-images.githubusercontent.com/84128937/121287897-6b845000-c91d-11eb-81bc-e291763a4168.png)

Fgrain is similar to the grain growth model in MOOSE, with an interaction term for phase and grain interaction, given as: 
![image](https://user-images.githubusercontent.com/84128937/121288124-d0d84100-c91d-11eb-9c3d-a7d8aee814c6.png)

Fphase is differentiates between phases and is given by: 
![image](https://user-images.githubusercontent.com/84128937/121288212-f6654a80-c91d-11eb-9b45-418e82405389.png)

Where phi(tau) is a function dependent on the temporal changes in the system i.e. tau = T/Tl; Tl is the liquidus temperature and T is the temperature at a given position and time.

My question is how do I implement the first two equations in MOOSE? They should be coupled together? No?
I understand that I can take in the temperature using a SolutionFunction or a SolutionUO or use a MultiApp (but the mesh will need to be the same, which I don't want).

Do I need to code the equations myself or is there a MOOSE system that I can utilize?

Regards,
Ali

**Reply by laagesen** [reply]
The first two equations are Allen-Cahn equations and you should be able to implement them with existing phase-field kernels. Please have a look at the phase-field documentation page:
https://mooseframework.inl.gov/modules/phase_field/index.html
And in particular the page on basic phase-field equations will explain the basic idea of how to use existing kernels to implement the Allen-Cahn equation:
https://mooseframework.inl.gov/modules/phase_field/Phase_Field_Equations.html
If you are wanting to simulate a large number of grains, it may be more convenient in the long run to write an action that will add all the necessary kernels. In this case, as in the current MOOSE grain growth model, you would not be using the free energy based approach but instead deriving the evolution equations explicitly and adding the corresponding kernels. More information about the grain growth model can be found here:
https://mooseframework.inl.gov/modules/phase_field/Grain_Growth_Model.html

**Reply by alikhan-zada** [reply]
Thank you Larry,
So I will have to write my own Kernel and implement an action system like in the Grain Growth Model?
Also the equations are coupled together as can be seen in the local free energy. They also have derivatives w.r.t. different variables.

Ali

**Reply by laagesen** [reply]
To try to answer your question as directly as possible, you can probably avoid writing any new kernels or actions if you don't mind having a very complex/long input file. This is totally fine and will not cause any problems, it just may be harder to read. To do this, you would create a DerivativeParsedMaterial that includes f_phase and f_grain. Then you would need to add kernels to create an Allen-Cahn equation for each variable as described in 
https://mooseframework.inl.gov/modules/phase_field/Phase_Field_Equations.html
The AllenCahn kernel for each variable will take the DerivativeParsedMaterial as an input parameter and will automatically take the partial derivative of the free energy with respect to that variable. You do not need to do anything special because of the fact that the variables are coupled together in the free energy; taking the partial derivative wrt each variable is all you need to do and this is handled automatically. ACInterface kernels will also be needed for the terms resulting from f_grad.

If you want to have simpler input files, you would need to create a new action that would handle the coupling terms in f_grain. You could probably use an AllenCahn kernel for d(f_phase)/d(zeta).

**Reply by alikhan-zada** [reply]
Hello Larry,
Thank you for your help so far.
For d(f_phase)/d(zeta) I have the following input file. Can you please take a look and let me know if what I have done so far is correct? 

Sorry for being such a bother 😅

[phase_orig.i.txt](https://github.com/idaholab/moose/files/6637432/phase_orig.i.txt)
